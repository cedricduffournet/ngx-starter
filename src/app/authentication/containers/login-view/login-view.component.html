<app-auth-layout
  title="{{ 'HELLO' | translate }},"
  title2="{{ 'AUTHENTICATE' | translate }}"
>
  <app-login-form
    [processing]="processing$ | async"
    [error]="error$ | async"
    (login)="onLogin($event)"
  ></app-login-form>
</app-auth-layout>
