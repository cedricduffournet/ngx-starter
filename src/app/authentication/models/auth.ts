export interface AuthToken {
  access_token: string;
  expires_in: number;
  token_type: string;
  refresh_token: string;
}

export interface Credentials {
  login: string;
  password: string;
}

export interface ResetPassword {
  plainPassword: {
    first: string;
    second: string;
  };
}

export interface Register extends ResetPassword {
  civility: number;
  firstname: string;
  lastname: string;
  email: string;
  confirmationUrl: string;
}

export interface SendResetPassword {
  username: string;
}
