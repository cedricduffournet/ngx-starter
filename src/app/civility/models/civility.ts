import { schema } from 'normalizr';

export interface Civility {
  id: number;
  name: string;
  code: string;
}

export const civilitySchema = new schema.Entity('civilities');
