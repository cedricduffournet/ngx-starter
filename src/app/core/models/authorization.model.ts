export interface Authorization {
  create: boolean;
  update: boolean;
  delete: boolean;
}
