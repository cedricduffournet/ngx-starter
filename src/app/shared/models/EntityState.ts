export interface EntityState<T> {
  [key: string]: T;
}
