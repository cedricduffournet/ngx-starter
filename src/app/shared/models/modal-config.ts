import { ModalOptions } from 'ngx-bootstrap/modal/modal-options.class';

export const GLOBAL_MODAL_CONFIG: ModalOptions = {
  animated: false
};

export const CRUD_MODAL_CONFIG: ModalOptions = {
  ...GLOBAL_MODAL_CONFIG,
  keyboard: false,
  ignoreBackdropClick: true
};
