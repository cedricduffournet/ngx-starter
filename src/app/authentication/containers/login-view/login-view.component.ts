import {
  ChangeDetectionStrategy,
  Component,
  OnInit,
  OnDestroy
} from '@angular/core';

import { Observable } from 'rxjs';
import { Store, select } from '@ngrx/store';
import { BsModalService } from 'ngx-bootstrap/modal';

import { Credentials } from '@app/authentication/models/auth';
import { LoginViewActions } from '@app/authentication/state/actions';
import * as fromAuth from '@app/authentication/state/reducers';

@Component({
  selector: 'app-login-view',
  templateUrl: './login-view.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class LoginViewComponent implements OnInit, OnDestroy {
  processing$: Observable<boolean>;
  error$: Observable<boolean>;

  constructor(
    private store: Store<fromAuth.State>,
    private modalService: BsModalService
  ) {}

  ngOnInit() {
    // force close all modal, for example user can be updating an entity in modal
    // and be redirect to login page, in this case modal is not hiden
    this.closeAllModals();
    this.processing$ = this.store.pipe(select(fromAuth.getLoginViewProcessing));
    this.error$ = this.store.pipe(select(fromAuth.getLoginViewError));
  }

  onLogin(credentials: Credentials) {
    this.store.dispatch(LoginViewActions.login({ credentials }));
  }

  ngOnDestroy() {
    this.store.dispatch(LoginViewActions.clear());
  }

  private closeAllModals() {
    for (let i = 1; i <= this.modalService.getModalsCount(); i++) {
      this.modalService.hide(i);
    }
  }
}
