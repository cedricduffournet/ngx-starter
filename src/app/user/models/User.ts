import { Civility } from '@app/civility/models/civility';

export interface UserGroup {
  name: string;
  roles: string[];
  superAdmin: boolean;
  admin: boolean;
}

export interface User {
  id: number;
  firstname: string;
  lastname: string;
  email: string;
  civility: Civility;
  roles: string[];
  groups: UserGroup[];
  enabled: boolean;
}
