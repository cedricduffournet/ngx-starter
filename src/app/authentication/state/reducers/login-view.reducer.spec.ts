import { reducer } from '@app/authentication/state/reducers/login-view.reducer';
import * as fromLoginView from '@app/authentication/state/reducers/login-view.reducer';
import { User } from '@app/user/models/User';
import { AuthToken } from '@app/authentication/models/auth';
import {
  LoginViewActions,
  AuthApiActions
} from '@app/authentication/state/actions';

describe('LoginViewReducer ', () => {
  describe('undefined action', () => {
    it('should return the default state', () => {
      const action = {} as any;

      const result = reducer(undefined, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('CLEAR', () => {
    it('should return default state on clear', () => {
      const initialState = {
        processing: true,
        error: true
      };
      const action = LoginViewActions.clear();
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });
  describe('LOGIN', () => {
    it('should set processing to true', () => {
      const credentials = {
        login: 'login',
        password: 'password'
      };
      const action = LoginViewActions.login({ credentials });
      const result = reducer(fromLoginView.INITIAL_STATE, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('LOGIN_SUCCESS', () => {
    it('should set processing to false', () => {
      const authToken = {
        access_token: 'token'
      } as AuthToken;

      const initialState = {
        processing: true
      } as fromLoginView.State;

      const action = AuthApiActions.loginSuccess({ authToken });
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('LOAD_LOGGED_USER_SUCCESS', () => {
    it('should set processing to false', () => {
      const user = {
        firstname: 'user'
      } as User;

      const initialState = {
        processing: true
      } as fromLoginView.State;

      const action = AuthApiActions.loadLoggedUserSuccess({ user });
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('LOGIN_FAILURE & LOAD_LOGGED_USER_FAILURE', () => {
    function authenticationFailed(
      action:
        | typeof AuthApiActions.loadLoggedUserFailure
        | typeof AuthApiActions.loginFailure,
      initialState: any
    ) {
      const createAction = action({ error: true });
      const result = reducer(initialState, createAction);
      expect(result).toMatchSnapshot();
    }

    it('should set processing to false and error to true ', () => {
      const initialState = {
        processing: true,
        error: false
      } as fromLoginView.State;

      authenticationFailed(AuthApiActions.loginFailure, initialState);
      authenticationFailed(AuthApiActions.loadLoggedUserFailure, initialState);
    });
  });
});
