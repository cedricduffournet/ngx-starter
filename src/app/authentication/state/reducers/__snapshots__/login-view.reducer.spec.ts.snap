// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginViewReducer  CLEAR should return default state on clear 1`] = `
Object {
  "error": false,
  "processing": false,
}
`;

exports[`LoginViewReducer  LOAD_LOGGED_USER_SUCCESS should set processing to false 1`] = `
Object {
  "error": false,
  "processing": false,
}
`;

exports[`LoginViewReducer  LOGIN should set processing to true 1`] = `
Object {
  "error": false,
  "processing": true,
}
`;

exports[`LoginViewReducer  LOGIN_FAILURE & LOAD_LOGGED_USER_FAILURE should set processing to false and error to true  1`] = `
Object {
  "error": true,
  "processing": false,
}
`;

exports[`LoginViewReducer  LOGIN_FAILURE & LOAD_LOGGED_USER_FAILURE should set processing to false and error to true  2`] = `
Object {
  "error": true,
  "processing": false,
}
`;

exports[`LoginViewReducer  LOGIN_SUCCESS should set processing to false 1`] = `
Object {
  "error": false,
  "processing": false,
}
`;

exports[`LoginViewReducer  undefined action should return the default state 1`] = `
Object {
  "error": false,
  "processing": false,
}
`;
